### PR TITLE
Mixed precision: fms_mod

### PR DESCRIPTION
**Pull request by J-Lentz**
**Description**
`monotonic_array` in `fms_mod` has been converted to mixed precision, and a unit test has been written for it.

**How Has This Been Tested?**
Unit test compiles, runs, and passes on the AMD box with gcc and Intel.

**Checklist:**
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [x] Any dependent changes have been merged and published in downstream modules
- [x] New check tests, if applicable, are included
- [x] `make distcheck` passes

### CODE REVIEW

File: fms/include/fms.inc [hunk 1]
```diff
@@ -16,729 +16,25 @@
...
+function MONOTONIC_ARRAY_(array, direction)
+real(FMS_MOD_KIND_), intent(in) :: array(:) !< An array of real values. If the size(array) < 2 this function
+                                            !! assumes the array is not monotonic, no fatal error will occur.
 integer, intent(out), optional :: direction !< If the input array is:
                                             !! >> monotonic (small to large) then direction = +1.
                                             !! >> monotonic (large to small) then direction = -1.
                                             !! >> not monotonic then direction = 0.
-logical :: monotonic_array !< If the input array of real values either increases or decreases monotonically
-                           !! then TRUE is returned, otherwise FALSE is returned.
+logical :: MONOTONIC_ARRAY_ !< If the input array of real values either increases or decreases monotonically
```
Comment (thomas-robinson) [inline fms/include/fms.inc:32]: I realize you didn't write this code @J-Lentz , but this seems very odd to me.  I would just declare this function as logical, or have a return value and declare that return value as logical.  This is somehow both and neither at the same time.

File: string_utils/fms_string_utils.F90 [hunk 2]
```diff
@@ -112,11 +114,20 @@ subroutine c_free(ptr) bind(c,name="free")
...
-!> Converts a number to a string
+!> Converts a number or a Boolean value to a string
 !> @ingroup fms_mod
 interface string
-   module procedure string_from_integer
-   module procedure string_from_real
+  module procedure string_from_logical
+  module procedure string_from_integer
+  module procedure string_from_r4, string_from_r8
+end interface
```
Comment (thomas-robinson) [inline string_utils/fms_string_utils.F90:123]: Will the conflict with 2023.01-beta3?

Comment (J-Lentz) [reply string_utils/fms_string_utils.F90]: Updated `fms_string_utils_mod` merged from main in 42c2d1c1caa18fb9854319c541a4b315ffa61c64.

File: fms/include/fms.inc (hunk 1 above)
Comment (J-Lentz) [reply fms/include/fms.inc]: Return variable has been replaced with `ret` in c245ad67f1778eca3a7543cd82cdb57db54a426e.

File: CMakeLists.txt [hunk 3]
```diff
@@ -288,9 +288,12 @@ foreach(kind ${kinds})
 
   target_include_directories(${libTgt}_f PRIVATE include
                                                  fms
+                                                 fms/include
                                                  fms2_io/include
+                                                 string_utils/include
                                                  mpp/include
-                                                 constants)
+                                                 constants
+                                                 test_fms/fms/include)
```
Comment (mlee03) [inline CMakeLists.txt:296]: not sure if test_fms is  needed for CMake yet. @rem1776, is this correct?

File: CMakeLists.txt [hunk 4]
```diff
@@ -324,8 +327,11 @@ foreach(kind ${kinds})
   target_include_directories(${libTgt} PUBLIC
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/include>
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/fms>
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/fms/include>
     $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/fms2_io/include>
-    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/mpp/include>)
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/string_utils/include>
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/mpp/include>
+    $<BUILD_INTERFACE:${CMAKE_CURRENT_SOURCE_DIR}/test_fms/fms/include>)
```
Comment (mlee03) [inline CMakeLists.txt:334]: here too

File: fms/Makefile.am [hunk 5]
```diff
@@ -32,6 +32,9 @@ noinst_LTLIBRARIES = libfms.la
 # Each convenience library depends on its source.
 libfms_la_SOURCES = \
   fms.F90 \
+  include/fms.inc \
```
Comment (mlee03) [inline fms/Makefile.am:35]: are these also needed for fms_mod.$(FC_MODEXT)?

File: fms/include/fms.inc [hunk 6]
```diff
@@ -16,729 +16,26 @@
...
 
 !> @brief Determines if a real input array has monotonically increasing or
 !!     decreasing values.
 !! @return If the input array of real values either increases or decreases monotonically then true
 !! is returned, otherwise false is returned.
-function monotonic_array ( array, direction )
-real,    intent(in)            :: array(:) !< An array of real values. If the size(array) < 2 this function
-                                           !! assumes the array is not monotonic, no fatal error will occur.
+
+function MONOTONIC_ARRAY_(array, direction) result(ret)
```
Comment (mlee03) [inline fms/include/fms.inc:25]: I guess this routine doesn't take into account elements having the same value, e.g. -2, -1, 0, 0, 1, 2

Comment (mlee03) [reply fms/include/fms.inc:25]: Need to get my math concepts straight.  Monotonic increasing/decreasing = strictly increasing/decreasing and not remaining constant. All's good.

File: CMakeLists.txt (hunk 3 above)
Comment (rem1776) [reply CMakeLists.txt:296]: Yeah that's correct we don't have cmake set up to run the tests so we don't need to compile them

File: fms/include/fms.inc (hunk 6 above)
Comment (J-Lentz) [reply fms/include/fms.inc:25]: I believe monotonic functions, as commonly defined, can have plateaus of constant value. So this function seems to use a peculiar definition of "monotonic".

Comment (bensonr) [reply fms/include/fms.inc:25]: There is a difference between a monotonic and a  _strictly_ monotonic function.  If this function only enforces (X<sub>1</sub> &lt; X<sub>2</sub>), then it is testing for _strict_ increasing monotonicity.  If it is (X<sub>1</sub> &gt; X<sub>2</sub>), then it is enforcing _strict_ decreasing monotonicity.

Comment (mlee03) [reply fms/include/fms.inc:25]: Perhaps it's worth commenting in the code that monotonic refers to strictly monotonic.

File: fms/include/fms_r4.fh
Comment (mlee03) [inline fms/include/fms_r4.fh:1]: #def #undef #def #undef include format

File: fms/include/fms_r8.fh
Comment (mlee03) [inline fms/include/fms_r8.fh:1]: here too

File: fms/include/fms.inc (hunk 6 above)
Comment (J-Lentz) [reply fms/include/fms.inc:25]: 0b397a624e7b5dd2ea825372a8f1ff66029b63fc

File: fms/include/fms_r4.fh
Comment (J-Lentz) [reply fms/include/fms_r4.fh:1]: 2f60d9b76757a959061f96fc69fe1ea205721280

File: fms/include/fms_r8.fh
Comment (J-Lentz) [reply fms/include/fms_r8.fh:1]: 2f60d9b76757a959061f96fc69fe1ea205721280

**Review by thomas-robinson — APPROVED** [review]

File: fms/Makefile.am (hunk 5 above)
Comment (J-Lentz) [reply fms/Makefile.am:35]: 850d957da3c0e28240215288b4de1aefdc05c132

**Pull request merged**